src/mlpack/tests/main_tests/hmm_training_tests.cpp

```diff
@@ -0,0 +1,209 @@
+/**
+ * @file hmm_training_tests.cpp
+ * @author Daivik Nema
+ *
+ * Test mlpackMain() of hmm_train_main.cpp.
+ */
+#include <string>
+#include <fstream>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "HMMTrain";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include "test_helper.hpp"
+#include <mlpack/methods/hmm/hmm_train_main.cpp>
+#include <mlpack/methods/hmm/hmm_model.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+#include <boost/archive/text_oarchive.hpp>
+#include <boost/archive/text_iarchive.hpp>
+
+using namespace mlpack;
+
+struct HMMTrainMainTestFixture
+{
+ public:
+  HMMTrainMainTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~HMMTrainMainTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(HMMTrainMainTest, HMMTrainMainTestFixture);
+
+inline void fileExists(std::string fileName)
+{
+  ifstream ifp(fileName);
+  if (!ifp.good())
+    BOOST_FAIL("Bad stream " + fileName);
+  ifp.close();
+}
+
+// Make sure that the number of states cannot be negative
+BOOST_AUTO_TEST_CASE(HMMTrainStatesTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = -3;  // Invalid!
+  std::string hmmType = "discrete";
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that tolerance is non negative
+BOOST_AUTO_TEST_CASE(HMMTrainToleranceNonNegative)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "gaussian";
+  double tol = - 100;  // Invalid
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("tolerance", tol);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
```

Comment:
@PlantsAndBuildings `hmm_train_main.cpp` currently doesn't have a check for the value of tolerance, that's may be the reason why there is an error as `BOOST_REQUIRE_THROW()` is expecting to receive a runtime error. You may solve this by adding a check in for tolerance in ` hmm_train_main`.  It's your choice to do so don't feel obligated.
May be something like this could be done:
```
RequireParamValue<int>("tolerance", [](int x) { return x >= 0; }, true,
        "tolerance must be non-negative");
```

Reply:
@manish7294 . Thanks, I've added the check like you suggested.